### ISSUE TITLE
Improve UseShellExecute on Unix

### ISSUE THREAD

**Issue by fmj02** [dotnet/corefx#19956]
This is just a question because from what I read in google there seems to be a problem with it. Does it work/will it work in 2.0?

```c#
private bool OpenBrowser(string url)
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                Process.Start(new ProcessStartInfo("cmd", $"/c start {url.Replace("&", "^&")}") { CreateNoWindow = true });
                return true;
            }
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            {
                Process.Start("xdg-open", url);
                return true;
            }
            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            {
                Process.Start("open", url);
                return true;
            }
            return false;
        }
```

**Comment by danmosemsft** [comment]
yes, does it not for you?

**Comment by svick** [comment]
Could you clarify what exactly did you read and where?

**Comment by mellinoe** [comment]
Yes, it is supported.

One common gotcha is that [`UseShellExecute` is not supported](https://github.com/dotnet/corefx/blob/964b73bddbc784b33192b51556f0d594462b67e4/src/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.Windows.cs#L81). The property exists, but you cannot actually set it. The code snippet you have shared here is a way you can get around that limitation.

**Comment by stephentoub** [comment]
> The property exists, but you cannot actually set it

It works on Unix:
https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L229

For Windows, I was hoping we'd fix https://github.com/dotnet/corefx/issues/18949 for 2.0, but I guess it got bumped out?

**Comment by mellinoe** [comment]
Yeah, it did not get fixed for 2.0. Given how the code you just linked works, can we do the same thing on Windows (e.g. just call `cmd /c <exe> <args>`) ?

**Comment by danmosemsft** [comment]
Perhaps should have started a broader dialog on that one for 2.0. The usage is quite significant.
https://apisof.net/catalog/System.Diagnostics.ProcessStartInfo.UseShellExecute 

I assume the OneCore issue can be solved by simply catching MissingMethodException. And I believe we support STA threads as you're not on Nano. (Where I guess ShellExecuteEx isn't available anyway)

Worth proposing as a DCR @stephentoub @mellinoe @Priya91 ?

**Comment by fmj022** [comment]
@svick https://brockallen.com/2016/09/24/process-start-for-urls-on-net-core/

**Comment by danmosemsft** [comment]
This tracks improving Unix implementation to copy [Mono](https://github.com/mono/mono/blob/8df83179b63e256bcd476d6d7ddbbd50015b81d6/mono/metadata/w32process-unix.c#L2125-L2150) instead of just [launching with sh](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L229-L240).

**Comment by TSlivede** [comment]
Just my vote for
>improving Unix implementation to copy Mono instead of just launching with sh

From my comment at https://github.com/dotnet/corefx/issues/4720#issuecomment-325702122

>I'm aware, that on linux "shell" most often means "[command line shell](https://en.wikipedia.org/wiki/Shell_(computing)#Text_.28CLI.29_shells)" but `UseShellExecute` was never about that. Instead `UseShellExecute` allows the shell (on Windows: shell32.dll not cmd.exe) to do some intermediate stuff, like interpreting the `ProcessStartInfo.Verb` property and deciding whether the given path actually belongs to an application or if it's just the path to a file which should be opened by another application. Test app:

	using System;
	using System.Diagnostics;

	public class MainClass{
		public static int Main(string[] args){
			ProcessStartInfo startInfo = new ProcessStartInfo(args[0]);
			startInfo.UseShellExecute=true;
			Process.Start(startInfo);
			startInfo.UseShellExecute=false;
			Process.Start(startInfo);
			return 0;
		}
	}

>Passing `notepad.exe` as argument opens notepad twice, while passing `path\to\file.txt` opens notepad once and geves an error if `UseShellExecute` is false. Notice: In neither case any console window is opened and passing a cmd command like `dir` does not succeed.

**Comment by maryamariyan** [comment]
Question A) I can see that the intent of issue #19956 is to improve UseShellExecute by using "open" on osx and using "xdg-open", "gnome-open", etc. on linux. But does this also mean if we are using Process.Start with UseShellExecute: true (for actions other than opening file/urls) then the behavior should remain as it used to be before?

Question B) Is it wrong to expect that Process.Start should only be able to open urls and existing files successfully and that it is not responsible to do anything else?

**Comment by TSlivede** [comment]
>But does this also mean if we are using Process.Start with UseShellExecute: true (for actions other than opening file/urls) then the behavior should remain as it used to be before?

I don't think so, that's not what UseShellExecute is about. IMHO the reference for the behaviour of UseShellExecute should be Windows .Net Framework. And although `exit 42` works  in cmd.exe, you get an error for both testcases on Windows. This is correct, because UseShellExecute means *not* "use command line shell".

If somebody really needs the `sh -c` feature he/she could start an API-Proposal to add some Property like `UseCommandLineShell`, that does `sh -c` on *nix and `cmd /c` on Windows - but I don't really see the need for this.

**Comment by TSlivede** [comment]
One more (incorrecly interpreted) meaning of `UseShellExecute`: Open in new Terminal.

From https://github.com/dotnet/corefx/issues/4720#issuecomment-161130544 :
>Probably can be accomplished by storing list of known terminals var KnownShells = {"/usr/bin/terminal", "/usr/bin/gnome-terminal", "/usr/bin/knosole"} and then create a virtual tty listener for communication between .NET process and terminal. The execution would be spawning a child process with available terminal and arguments. This can also be restricted to desktop context in Unix as in Windows. Would be nice to also let user override the terminal choice using an environment variable say CORECLR_TERMINAL=/usr/bin/gnome-terminal in case system has multiple choices and the selection based on KnownShells list precedence is not what user prefers.

As I mentioned, I don't think, that's what `UseShellExecute` is about. However maybe a new property could be added to `ProcessStartInfo` via an API-Proposal? (maybe `ProcessStartInfo.OpenInTerminal` ?). If it is set to true, the new process is launched within a terminal on linux and something like [create new console](https://msdn.microsoft.com/en-us/library/windows/desktop/ms684863.aspx#create_new_console) happens on Windows.

<sup>I moved this comment from https://github.com/dotnet/corefx/issues/4720 , seem to fit better here.</sup>

**Comment by maryamariyan** [comment]
closed by #23705

**Comment by jnm2** [comment]
Speaking of UseShellExecute opening a new conhost, I believe this is the minimum amount of work necessary to use shell execute and stay in the same conhost window: https://github.com/jnm2/Elevate
